server/src/main/java/io/crate/operation/collect/files/CSVLineParser.java

```diff
@@ -56,25 +72,41 @@ public CSVLineParser(CopyFromParserProperties properties) {
 
     public void parseHeader(String header) throws IOException {
         MappingIterator<String> iterator = csvReader.readValues(header.getBytes(StandardCharsets.UTF_8));
-        iterator.readAll(keyList);
-        HashSet<String> keySet = new HashSet<>(keyList);
+        int i = 0;
+        while (iterator.hasNext()) {
+            String headerPart = iterator.next();
+            headerKeyList.add(headerPart);
+            if (targetColumns.contains(headerPart)) {
+                keyPositions.put(i, headerPart);
+            }
+            i++;
+        }
+        HashSet<String> keySet = new HashSet<>(headerKeyList);
         keySet.remove("");
-        if (keySet.size() != keyList.size() || keySet.size() == 0) {
+        if (keySet.size() != headerKeyList.size() || keySet.size() == 0) {
             throw new IllegalArgumentException("Invalid header: duplicate entries or no entries present");
         }
     }
 
     public byte[] parse(String row) throws IOException {
-        MappingIterator<Object> iterator = csvReader.readValues(row.getBytes(StandardCharsets.UTF_8));
+        MappingIterator<String> iterator = csvReader.readValues(row.getBytes(StandardCharsets.UTF_8));
         out.reset();
         XContentBuilder jsonBuilder = new XContentBuilder(JsonXContent.JSON_XCONTENT, out).startObject();
-        int i = 0;
-        while (iterator.hasNext()) {
-            if (i >= keyList.size()) {
-                throw new IllegalArgumentException("Number of values exceeds number of keys");
+        List<String> valueList = new ArrayList<>();
+        iterator.readAll(valueList);
+        if (headerExists && valueList.size() > headerKeyList.size()) {
+            throw new IllegalArgumentException("Number of values exceeds number of keys in csv file");
+        }
+        if (!headerExists && keyPositions.size() > valueList.size()) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "Expected %d values, encountered " +
+                                               "%d at line X. This is not allowed when there is no " +
+                                                "header provided)", keyPositions.size(), valueList.size()));
+        }
+        for (int i = 0; i < valueList.size(); i++) {
```

Comment:
I think this way, the `row` is iterated twice. (could have performance impact)

Reply:
@jeeminso I have optimized the parse method and there is now one iteration. I have also added a new check if there is a header and the target columns are covered to skip handling extra values. I mentioned it above in the pr description at the second issue.

Reply:
Thanks, maybe one more thing could be to take the two `if`s that throw exceptions out of the loop. I think this could help avoid conditionals and in my opinion performance is not so critical for failing scenarios if there is gain for passing cases.

Reply:
I took the 

1. 
```
if (!headerExists && (keyPositions.size() > i)) {
            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "Expected %d values, " +
                                               "encountered %d at line %d. This is not allowed when there " +
                                               "is no header provided)",keyPositions.size(), i, rowNumber));
}
```
out of the loop,

but the

2.
```
if (headerExists && i >= headerKeyList.size()) {
                throw new IllegalArgumentException("Number of values exceeds number of keys in csv file");
}
```
cannot be applied if there is this optimization of stopping the iteration if target columns have been covered

3. 
```
if (headerExists && keyPositions.size() == j) {
                break;
}
```
I am not sure if this optimization is necessary as it deals with with csv lines with many values and the user has chosen only a small part of the available columns.

Overall if we want to keep (3), (1) has to be inside loop